File: apex_dashboard/apex_dashboard/page/expense_dashboard/expense_dashboard.py
import frappe
from frappe import _
from frappe.utils import flt, getdate, add_months, add_days, get_first_day, get_last_day, today

@frappe.whitelist()
def get_dashboard_data(company=None, period="This Month", from_date=None, to_date=None):
    # Check cache first
    from apex_dashboard.cache_utils import get_cached_dashboard_data, set_dashboard_cache
    
    cache_key_suffix = f"{company}_{period}_{from_date}_{to_date}"
    cache_key = f"expense_{cache_key_suffix}"
    cached_data = frappe.cache().get_value(cache_key)
    if cached_data:
        return cached_data
    
    if not company:
        company = frappe.defaults.get_user_default("Company")

    # Use custom dates if provided, otherwise calculate from period
    if period == "Custom" and from_date and to_date:
        from_date = getdate(from_date)
        to_date = getdate(to_date)
    else:
        from_date, to_date = get_period_dates(period)
    
    # Get Company Currency
    currency = frappe.get_value("Company", company, "default_currency")

    # Fetch Config
    config = frappe.get_single("Apex Dashboard Config")
    
    categories = {}
    total_expense = 0.0
    
    # Fetch all relevant GL entries once (Optimization)
    gl_entries = frappe.db.sql("""
        SELECT account, debit
        FROM `tabGL Entry`
        WHERE company = %s 
            AND posting_date BETWEEN %s AND %s 
            AND is_cancelled = 0
            AND debit > 0
    """, (company, from_date, to_date), as_dict=1)

    if config.cards:
        for row in config.cards:
            # Fetch linked visual card details
            card_doc = frappe.get_doc("Apex Dashboard Card", row.card)
            
            if card_doc.category != "Expenses" or not card_doc.is_active:
                continue

            if row.card_source == "Static Value":
                static_value = flt(row.static_value)
                
                categories[card_doc.card_title] = {
                    'name': card_doc.card_title,
                    'total': static_value,
                    'color': card_doc.color,
                    'accounts': []
                }
                total_expense += static_value

            elif row.card_source == "Chart of Accounts" and row.account:
                parent_account = row.account
                
                if not parent_account:
                    continue

                # Fetch all child accounts under the parent
                if frappe.db.exists("Account", parent_account):
                    parent_lft, parent_rgt = frappe.db.get_value("Account", parent_account, ["lft", "rgt"])
                    
                    child_accounts = frappe.db.sql("""
                        SELECT name, account_name 
                        FROM `tabAccount` 
                        WHERE lft >= %s AND rgt <= %s AND is_group = 0 AND docstatus = 0
                    """, (parent_lft, parent_rgt), as_dict=1)
                    
                    card_total = 0.0
                    account_details = []
                    
                    for account in child_accounts:
                        # Filter GL Entries for this account
                        account_gl_entries = [
                            gl for gl in gl_entries 
                            if gl.account == account['name']
                        ]
                        
                        expense_value = sum(gl['debit'] for gl in account_gl_entries) if account_gl_entries else 0.0
                        
                        if expense_value > 0:
                            card_total += expense_value
                            account_details.append({
                                'account_name': account['account_name'],
                                'amount': expense_value
                            })
                    
                    if card_total > 0:
                        account_details.sort(key=lambda x: x['amount'], reverse=True)
                        
                        # Check if card already exists in categories (from previous row)
                        if card_doc.card_title in categories:
                            # Merge with existing data
                            categories[card_doc.card_title]['total'] += card_total
                            categories[card_doc.card_title]['accounts'].extend([
                                {'name': d['account_name'], 'balance': d['amount']} for d in account_details
                            ])
                            # Re-sort accounts by balance
                            categories[card_doc.card_title]['accounts'].sort(key=lambda x: x['balance'], reverse=True)
                        else:
                            # Create new entry
                            categories[card_doc.card_title] = {
                                'name': card_doc.card_title,
                                'total': card_total,
                                'color': card_doc.color, # Use hex color directly
                                'accounts': [
                                    {'name': d['account_name'], 'balance': d['amount']} for d in account_details
                                ]
                            }
                        total_expense += card_total

    # Convert dict to list and sort by total
    category_list = list(categories.values())
    category_list.sort(key=lambda x: x["total"], reverse=True)

    data = {
        'groups': category_list,
        'total_expense': total_expense,
        'currency': currency,
        'period': period,
        'from_date': str(from_date),
        'to_date': str(to_date)
    }
    
    # Cache for 5 minutes
    frappe.cache().set_value(cache_key, data, expires_in_sec=300)
    
    return data

def get_period_dates(period):
    current_date = getdate(today())
    
    if period == "Today":
        from_date = current_date
        to_date = current_date
    elif period == "This Week":
        # Week starts on Sunday (or Monday depending on locale, but let's assume standard start)
        # frappe.utils.get_first_day_of_week returns Monday usually
        # Let's use relative delta to find start of week
        from_date = add_days(current_date, -current_date.weekday()) # Monday
        to_date = add_days(from_date, 6) # Sunday
    elif period == "This Month":
        from_date = get_first_day(current_date)
        to_date = get_last_day(current_date)
    elif period == "Last Month":
        last_month = add_months(current_date, -1)
        from_date = get_first_day(last_month)
        to_date = get_last_day(last_month)
    elif period == "This Year":
        from_date = getdate(f"{current_date.year}-01-01")
        to_date = getdate(f"{current_date.year}-12-31")
    elif period == "Last Year":
        from_date = getdate(f"{current_date.year - 1}-01-01")
        to_date = getdate(f"{current_date.year - 1}-12-31")
    elif period == "All Time":
        from_date = getdate("1900-01-01")
        to_date = getdate("2099-12-31")
    else:
        # Default to this month
        from_date = get_first_day(current_date)
        to_date = get_last_day(current_date)
    
    return from_date, to_date

File: apex_dashboard/apex_dashboard/page/expense_dashboard/expense_dashboard.js
frappe.pages['expense_dashboard'].on_page_load = function (wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Expense Dashboard',
		single_column: true
	});

	// Load CSS
	frappe.require('/assets/apex_dashboard/page/expense_dashboard/expense_dashboard.css');

	const TEMPLATE = `
<div class="expense-dashboard">
    <div class="dashboard-header">
        <div class="header-content">
            <h1>Expense Overview</h1>
            <p class="subtitle">Track your company expenses</p>
        </div>
        <div class="header-actions">
            <div class="total-expense-card">
                <span class="label">Total Expenses</span>
                <span class="value" id="total-expense-value">Loading...</span>
            </div>
            <button class="btn btn-glass" id="config-btn" title="Configure Dashboard">
                <i class="fa fa-cog"></i>
            </button>
            <button class="btn btn-glass" id="refresh-btn">
                <i class="fa fa-refresh"></i>
            </button>
        </div>
    </div>

    <div id="dashboard-loader" class="loader-container">
        <div class="spinner"></div>
        <p>Fetching expense data...</p>
    </div>

    <div id="dashboard-content" class="dashboard-grid" style="display: none;">
        <!-- Category Cards will be injected here -->
    </div>

    <div class="dashboard-footer">
        <p>Data based on GL Entries. Last updated: <span id="last-updated">-</span></p>
    </div>
</div>
    `;

	// Load HTML Template
	$(TEMPLATE).appendTo(page.body);

	// Initialize
	new ExpenseDashboard(wrapper, page);
}

class ExpenseDashboard {
	constructor(wrapper, page) {
		this.wrapper = $(wrapper);
		this.page = page;
		this.format_currency = this.format_currency.bind(this);

		this.setup_filters();
		this.bind_events();
		this.fetch_data();
	}

	setup_filters() {
		this.company_field = this.page.add_field({
			fieldname: 'company',
			label: __('Company'),
			fieldtype: 'Link',
			options: 'Company',
			default: frappe.defaults.get_user_default("Company"),
			change: () => this.fetch_data()
		});

		this.period_field = this.page.add_field({
			fieldname: 'period',
			label: __('Period'),
			fieldtype: 'Select',
			options: [
				{ label: __('Today'), value: 'Today' },
				{ label: __('This Week'), value: 'This Week' },
				{ label: __('This Month'), value: 'This Month' },
				{ label: __('Last Month'), value: 'Last Month' },
				{ label: __('This Year'), value: 'This Year' },
				{ label: __('Last Year'), value: 'Last Year' },
				{ label: __('All Time'), value: 'All Time' },
				{ label: __('Custom'), value: 'Custom' }
			],
			default: 'All Time',
			change: () => {
				this.toggle_custom_dates();
				this.fetch_data();
			}
		});

		// Custom date fields (hidden by default)
		this.from_date_field = this.page.add_field({
			fieldname: 'from_date',
			label: __('From Date'),
			fieldtype: 'Date',
			default: frappe.datetime.month_start(),
			change: () => this.fetch_data()
		});

		this.to_date_field = this.page.add_field({
			fieldname: 'to_date',
			label: __('To Date'),
			fieldtype: 'Date',
			default: frappe.datetime.month_end(),
			change: () => this.fetch_data()
		});

		// Hide custom dates initially
		this.toggle_custom_dates();
	}

	toggle_custom_dates() {
		const period = this.period_field.get_value();
		const isCustom = period === 'Custom';

		this.from_date_field.df.hidden = !isCustom;
		this.to_date_field.df.hidden = !isCustom;
		this.from_date_field.refresh();
		this.to_date_field.refresh();
	}

	bind_events() {
		this.wrapper.find('#refresh-btn').on('click', () => {
			this.fetch_data();
		});

		// Config Button
		this.wrapper.find('#config-btn').on('click', () => {
			frappe.set_route('Form', 'Apex Dashboard Config', 'Apex Dashboard Config');
		});

		// Back to Hub
		this.wrapper.find('.header-content').prepend(`
            <button class="btn-glass" id="back-btn">
                <i class="fa fa-arrow-left"></i> Hub
            </button>
        `);

		this.wrapper.find('#back-btn').on('click', () => {
			frappe.set_route('apex_dashboards');
		});
	}

	fetch_data() {
		console.log("Expense Dashboard: Starting to fetch data...");
		this.show_loader(true);

		const company = this.company_field.get_value();
		const period = this.period_field.get_value();

		const args = {
			company: company,
			period: period
		};

		// Add custom dates if period is Custom
		if (period === 'Custom') {
			args.from_date = this.from_date_field.get_value();
			args.to_date = this.to_date_field.get_value();
		}

		console.log("Expense Dashboard: Company =", company, "Period =", period);

		frappe.call({
			method: 'apex_dashboard.apex_dashboard.page.expense_dashboard.expense_dashboard.get_dashboard_data',
			args: args,
			callback: (r) => {
				console.log("Expense Dashboard: API Response received", r);
				this.show_loader(false);
				if (r.message) {
					this.render(r.message);
				} else {
					console.error("Expense Dashboard: No data returned");
					this.wrapper.find('#dashboard-content').html(
						'<div class="alert alert-danger">Error loading dashboard data.</div>'
					);
				}
			},
			error: (r) => {
				this.show_loader(false);
				console.error("Expense Dashboard: API Error", r);
				frappe.msgprint({
					title: __('Error'),
					indicator: 'red',
					message: __('Failed to load dashboard data.')
				});
			}
		});
	}

	show_loader(show) {
		if (show) {
			this.wrapper.find('#dashboard-loader').show();
			this.wrapper.find('#dashboard-content').hide();
		} else {
			this.wrapper.find('#dashboard-loader').hide();
			this.wrapper.find('#dashboard-content').fadeIn();
		}
	}

	render(data) {
		console.log("Expense Dashboard: Rendering data", data);

		// Update Header
		try {
			const totalExpense = data.total_expenses || 0;
			const currency = data.currency || 'EGP';
			this.wrapper.find('#total-expense-value').text(this.format_currency(totalExpense, currency));
			this.wrapper.find('#last-updated').text(frappe.datetime.now_datetime());
		} catch (e) {
			console.error("Expense Dashboard: Error updating header", e);
		}

		// Render Categories
		const grid = this.wrapper.find('#dashboard-content');
		grid.empty();

		if (!data.groups || data.groups.length === 0) {
			grid.html('<div class="alert alert-info">No expenses found for this period.</div>');
			return;
		}

		data.groups.forEach(group => {
			// Use color from backend for gradient background and border
			let cardStyle = '';
			if (group.color) {
				// Gradient background (lighter) + darker border
				cardStyle = `
                    
				background: linear-gradient(135deg, ${group.color}40 0%, ${group.color}10 100%);
				
                    border-left: 4px solid ${group.color} !important;
			`;
			}

			let accountsHtml = '';
			if (group.accounts && group.accounts.length > 0) {
				group.accounts.forEach(acc => {
					accountsHtml += `
                        <div class="account-row">
                            <div class="account-info">
                                <span class="account-name">${acc.name}</span>
                            </div>
                            <span class="account-balance">${this.format_currency(acc.balance, data.currency)}</span>
                        </div>
                    `;
				});
			}

			const card = $(`
                <div class="category-card" style="${cardStyle}">
                    <div class="category-header">
                        <span class="expand-icon">▼</span>
                        <span class="category-name">${group.name}</span>
                        <span class="category-total">${this.format_currency(group.total, data.currency)}</span>
                    </div>
                    <div class="accounts-list" style="display: none;">
                        ${accountsHtml}
                    </div>
                </div>
            `);

			// Accordion Logic
			card.find('.category-header').on('click', function () {
				const body = $(this).next('.accounts-list');
				const icon = $(this).find('.expand-icon');

				body.slideToggle(200);
				icon.toggleClass('expanded');
			});

			grid.append(card);
		});
	}

	format_currency(amount, currency, precision = 2) {
		return new Intl.NumberFormat('en-US', {
			style: 'currency',
			currency: currency,
			minimumFractionDigits: precision,
			maximumFractionDigits: precision
		}).format(amount);
	}

	getContrastColor(hexColor) {
		// Convert hex to RGB
		if (!hexColor) return '#ffffff';

		// Remove # if present
		hexColor = hexColor.replace('#', '');

		// Convert to RGB
		const r = parseInt(hexColor.substr(0, 2), 16);
		const g = parseInt(hexColor.substr(2, 2), 16);
		const b = parseInt(hexColor.substr(4, 2), 16);

		// Calculate luminance
		const luminance = (0.299 * r + 0.587 * g + 0.114 * b) / 255;

		// Return black for light colors, white for dark colors
		return luminance > 0.5 ? '#000000' : '#ffffff';
	}

	get_theme_class(name) {
		const themes = ['theme-blue', 'theme-green', 'theme-orange', 'theme-purple', 'theme-pink'];
		const hash = name.split('').reduce((acc, char) => acc + char.charCodeAt(0), 0);
		return themes[hash % themes.length];
	}

	get_theme_class_from_name(themeName) {
		// Map Config color names to CSS classes
		const themeMap = {
			'Blue': 'theme-blue',
			'Green': 'theme-green',
			'Red': 'theme-red',
			'Purple': 'theme-purple',
			'Orange': 'theme-gold',  // Using gold as closest to orange
			'Teal': 'theme-lime'     // Using lime as closest to teal
		};
		return themeMap[themeName] || 'theme-blue';
	}
}
